src/adapter/src/catalog.rs

```diff
@@ -2517,12 +2515,25 @@ pub enum StorageSinkConnectionState {
 #[derive(Debug, Clone, Serialize)]
 pub struct View {
     pub create_sql: String,
-    pub optimized_expr: OptimizedMirRelationExpr,
+    pub expr: MirRelationExpr,
+    #[serde(skip_serializing)]
+    pub o_expr: OnceLock<OptimizedMirRelationExpr>,
     pub desc: RelationDesc,
     pub conn_id: Option<ConnectionId>,
     pub resolved_ids: ResolvedIds,
 }
 
+impl View {
+    pub fn optimized_expr(&self) -> &OptimizedMirRelationExpr {
```

Comment:
I cannot think of a better solution than widening the stored type to 

```rust
Result<OptimizedMirRelationExpr, TransformError>
```

and calling `get_or_init(...)` with `?` in client code.